qiime2/sdk/actiongraph.py

```diff
@@ -0,0 +1,215 @@
+        Delineates if getting the action input, param, or output types
+
+    Returns
+    -------
+    List of tuples containing name and required semantic types
+    List of tuples containing name and optional semantic types
+    """
+
+    req = []
+    non_req = []
+
+    if type == "input":
+        for k, v in action.signature.inputs.items():
+            if not v.has_default():
+                req.append([k, v.qiime_type])
+            else:
+                non_req.append(["."+k, [v.qiime_type, None]])
+    elif type == "param":
+        for k, v in action.signature.parameters.items():
+            if not v.has_default():
+                req.append([k, v.qiime_type])
+            else:
+                non_req.append(["."+k, [v.qiime_type, None]])
+    else:
+        for k, v in action.signature.outputs.items():
+            if not v.has_default():
+                req.append([k, v.qiime_type])
+            else:
+                non_req.append(["."+k, [v.qiime_type, None]])
+
+    return req, non_req
+
+
+def unravel(l):
+    """
+    Unravel Union node to get all permutations of types for each action
+
+    Parameters
+    ----------
+    list : list of Qiime2.types
+
+    Returns
+    -------
+    list of lists - list of permuations of types for each action
+
+    """
+
+    result = [l]
+    for i, x in enumerate(l):
+        if len(list(x[1])) > 1:
+            members = list(x[1])
+            temp = copy.deepcopy(result)
+
+            # update result with first element of types in member
+            for each_list in result:
+                each_list[i][1] = members[0]
+
+            # add in other permutations of types in member
+            for n in range(1, len(members)):
+                copy_result = copy.deepcopy(temp)
+                for each_list in copy_result:
+                    each_list[i][1] = members[n]
+                result += copy_result
+    return result
+
+
+def generate_nodes_by_action(action, opt):
+    """
+    Given a method, generates all combinations of inputs and
+    outputs for that particular method and and stores the combinations
+    as dictionaries in a resulting list.
+
+    Parameters
+    ----------
+    method : Qiime2.action
+    opt : {True, False}
+        Delineates if optional types should be included
+
+    Returns
+    -------
+    list of dictionaries - each dictionary is a combination inputs and
+        outputs for particular node
+
+    """
+
+    input, input_nr = get_next_arguments(action, "input")
+    param, param_nr = get_next_arguments(action, "param")
+    output, output_nr = get_next_arguments(action, "output")
+
+    input = unravel(input)
+    param = unravel(param)
+
+    opt_in_list = []
+
+    if opt:
+        opt_in_list += input_nr
+        opt_in_list += param_nr
+        opt_in_list = unravel(opt_in_list)
+        ins = [dict(x) for x in
+               [list(chain.from_iterable(i)) for i in
+                list(product(input, param, opt_in_list))]]
```

Comment:
This took me a few to track down, but I think there's an error here:

The product should be calculated on something like:
```python
product(*[  # the star here unpacks the list into N arguments for product
   [[key1, type1], [key1, type2]],  # maybe this is an input which was a union of type1/2
   [[key2, typex]],  # maybe these are params
   [[key3, typey]]
])
```
This is subtly different from what you have, which from unravel will create the product of only 3 lists, one of which might look like this:
```python
[  # this outer list is messing things up
[['table', FeatureTable[Frequency]],
  ['representative_sequences', FeatureData[Sequence]],
  ['stats', DeblurStats]]  # this inner list is saying that all of these are part of the same set to take the product of
                           # but we want to take the product of sets where each set is a single key/param name
]
```
Because your three lists typically contain a single list, the cartesian product is just the same 3 lists back out, which why the results _seem_ sensible at first glance.

Reply:
Update: I had misunderstood

input/param/opt_in_list are themselves cartesian products, so the product of a product is still the same product as the the product of the individual elements